CARDS-953: Number question Slider display type

acrowthe opened the pull request:
- Add the slider display type to decimal, long and double number questions.
- Includes Questionnaire Editor support
- Added a slider test form in the test runmode

acrowthe committed:
CARDS-953: Number question Slider display type
- Add the slider display type to decimal, long and double number questions.
- Includes Questionnaire Editor support

marta- committed:
CARDS-953/CARDS-1418: Enable the slider display mode in the questionnaire wizard
Only allow "horizontal" and "vertical" as values for sliderOrientation

marta- committed:
CARDS-953/CARDS-1419: Implement slider display in edit mode
Don't show the message "Please enter values between X and Y" when displayMode=slider, since the slider enforces the limits already

marta- committed:
CARDS-953/CARDS-1419: Implement slider display in edit mode
Updated documentation and propTypes

marta- committed:
CARDS-953/CARDS-1419: Implement slider display in edit mode
Specified the default sliderOrientation without string duplication

marta- committed:
CARDS-953/CARDS-1419: Implement slider display in edit mode
Bugfix: AnswerInstructions not displayed for mandatory questions with displayMode=slider and isRange=false

marta- committed:
CARDS-953/CARDS-1419: Implement slider display in edit mode
Some styling adjustments to prevent the selected value from covering the question text.
Also set a max-width to horizontal sliders.

marta- reviewed (approved):
Looks great

sashaandjic reviewed (approved):
tested
![image](https://user-images.githubusercontent.com/23243244/137568898-5813e5ca-4b60-4127-a506-787233167983.png)

marta- commented:
This can be merged as far as I am concerned even with the range bug. The values appear to be sent correctly to the server. What happens after should be fixed separately.

sashaandjic commented:
When we introduce slider, we should adjust the instruction text to.
![image](https://user-images.githubusercontent.com/23243244/137569100-714d4fd9-5149-4c24-9707-d79df0b9903b.png)

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -50,6 +50,12 @@
+          },
+          "slider": {
+            "isRange" : "boolean",
+            "sliderStep": "double",
+            "sliderMarkStep": "double",
```
Shouldn't these be integers? We're entering a `long` value, so a step of `0.5` does not make sense.

sdumitriu replied:
Yep, setting 0.5 for a long answer leads to unusable "left/down" keys not able to select the previous valid value, and setting `0.1` leads to none of the arrows working.

marta- replied:
Needs more extensive refactoring of the questionnaire wizard. Specifying `long` here won't prevent the user from entering 1.7, but will pop the login dialog when trying to save it due to server error (invalid value). This is not a blocker, can be done later.

sdumitriu reviewed (commented):
Bug: 
- create new Slider Test form
- select non-0 values for all sliders except for `Long slider, vertical`
- save and view-> form is incomplete (correct)
- edit -> `Long slider, vertical` is marked as mandatory and not set (correct)
- select a non-0 value for it, save and view -> form is complete (correct)
- edit, select 0 for `Long slider, vertical` -> question appears set (not red)
- save -> Form is still complete (correct)
- edit -> question is red again (wrong)
- save -> Form is invalid and incomplete (wrong), and the answer was removed

This highlights a problem with the implementation: `0` is considered as `answer not set`.

Bonus: for improved usability I would like to be able to be able to type numbers.

Just me? Tabbing through sliders is great except for vertical range sliders: first the bottom one is selected, then the top one, which makes sense logically (lower value first), but visually I am used for the tab to go down the page. Is anybody else bothered by this?

marta- committed:
CARDS-953/CARDS-1419: Implement slider display in edit mode
Bugfix: value 0 is mistaken for "no answer" on single sliders.

marta- committed:
CARDS-953/CARDS-1419: Implement slider display in edit mode
Bugfix: value 0 is mistaken for "no answer" on range sliders.

sashaandjic reviewed (approved):
tested

sashaandjic commented:
While testing this PR I realized that adding the form's status, especially in the view mode, would be beneficial. 
In order to see the status of the form I had to go back to the patient chart.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/NumberQuestion.jsx:
```diff
@@ -88,6 +116,37 @@ function NumberQuestion(props) {
+  let sliderMarks = [{value: minValue, label: minValue}, {value: maxValue, label: maxValue}];
+  if (typeof(sliderMarkStep) !== "undefined") {
+    let i = minValue + sliderMarkStep;
+    while (i < maxValue) {
```
vertical, minValue 0, maxValue 21, mark step 4 -> the labels for 20 and 21 are too close. I would not include the label for 20 in this case, just ...12, 16, 21.

```suggestion
    while (i <= maxValue - sliderMarkStep) {
```

sdumitriu committed:
CARDS-953: Number question Slider display type

Don't display a tickmark that's too close to the end if the mark step is not a divisor of the total.

sdumitriu commented:
Hum, one more possible (future) improvement? If the interval is not evenly divisible by the step size and is centered on 0, then the steps should also be centered on 0, e.g. `[-100..100] / 3` should have steps `[-100, -96, -93 ... -3, 0, 3, ... 93, 96, 100]` instead of `[-100, -97, -94 ... -1, 2, 5, ... 95, 100]`.
